azurelinuxagent/common/cgroupconfigurator.py

```diff
@@ -53,182 +75,229 @@ def initialize(self):
+            add_event(op=WALAEventOperation.CGroupsInfo, message=message)
 
-        def resource_limits_enforced(self):
-            return False
+        @staticmethod
+        def __log_cgroup_warning(format_string, *args):
+            message = format_string.format(*args)
+            logger.info(message)  # log as INFO for now, in the future it should be logged as WARNING
+            add_event(op=WALAEventOperation.CGroupsInfo, message=message, is_success=False, log_event=False)
 
-        def enable(self):
-            if not self._cgroups_supported:
-                raise CGroupsException("Attempted to enable cgroups, but they are not supported on the current platform")
+        def __check_no_legacy_cgroups(self):
+            """
+            Older versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent. When running
+            under systemd this could produce invalid resource usage data. Cgroups should not be enabled under this condition.
+            """
+            legacy_cgroups = self._cgroups_api.cleanup_legacy_cgroups()
+            if legacy_cgroups > 0:
+                self.__log_cgroup_warning("The daemon's PID was added to a legacy cgroup; will not monitor resource usage.")
+                return False
+            return True
 
-            self._cgroups_enabled = True
+        def __get_cgroup_controllers(self):
+            #
+            # check v1 controllers
+            #
+            cpu_controller_root, memory_controller_root = self._cgroups_api.get_cgroup_mount_points()
 
-        def disable(self):
-            self._cgroups_enabled = False
-            CGroupsTelemetry.reset()
+            if cpu_controller_root is not None:
+                logger.info("The CPU cgroup controller is mounted at {0}", cpu_controller_root)
+            else:
+                self.__log_cgroup_warning("The CPU cgroup controller is not mounted")
 
-        def create_slices(self):
-            if not self.enabled():
-                return
+            if memory_controller_root is not None:
+                logger.info("The memory cgroup controller is mounted at {0}", memory_controller_root)
+            else:
+                self.__log_cgroup_warning("The memory cgroup controller is not mounted")
 
-            # Create root slices for agent and agent and extensions for systemd-managed distros.
-            # The hierarchy is as follows:
-            # ├─user.slice
-            # ...
-            # ├─system.slice
-            # ...
-            # └─azure.slice
-            #   └─azure-vmextensions.slice
+            #
+            # check v2 controllers
+            #
+            cgroup2_mount_point, cgroup2_controllers = self._cgroups_api.get_cgroup2_controllers()
+            if cgroup2_mount_point is not None:
+                self.__log_cgroup_info("cgroups v2 mounted at {0}.  Controllers: [{1}]", cgroup2_mount_point, cgroup2_controllers)
```

Comment:
I know this is old code, but what are we doing with the v2 controllers? Is it just for informational purposes for now?

Reply:
It's for debugging purposes. Controllers cannot be mounted both on the v1 and v2 hierarchies, so when there is no v1 controller I want to know if it's because it's mounted in the v2 hierarchy. Telemetry shows that there is a very small number of vms mounting cpu/memory in v2.